### PR TITLE
Fix the #143392

### LINKED ISSUE #143392: Buttons to close tabs to right enabled even though there are no tabs to right.
Issue Type: <b>Bug</b>

This issue does not happen when only one tab is open in editor.

Steps to reproduce

1. Open two or more tabs
2. Right click on the right most tab header
3. You will see a menu "Close to the enabled right"

VS Code version: Code 1.64.2 (f80445acd5a3dadef24aa209168452a3d97cc326, 2022-02-09T22:00:58.347Z)
OS version: Darwin x64 21.3.0
Restricted Mode: No

### PR DESCRIPTION
**Pull request by Kalmaegi**
This PR fixes #143392

### CODE REVIEW

**Comment by weartist** [comment]
@bpasero There have been some unforeseen circumstances, I resubmitted it.

**Review by bpasero — CHANGES_REQUESTED** [review]
I would suggest to look closer at an existing example of a context key that is updated within a group and also for the context menu and just monkey-see-monkey-doo.

https://github.com/microsoft/vscode/blob/7c1c5300513617d79e1c66c8d2117a5b95259ffa/src/vs/workbench/common/contextkeys.ts#L40

The context key should not be specific for the context menu, because that is just a corner case.

**Comment by weartist** [comment]
> I would suggest to look closer at an existing example of a context key that is updated within a group and also for the context menu and just monkey-see-monkey-doo.
> 
> https://github.com/microsoft/vscode/blob/7c1c5300513617d79e1c66c8d2117a5b95259ffa/src/vs/workbench/common/contextkeys.ts#L40
> 
> The context key should not be specific for the context menu, because that is just a corner case.

I misunderstood the previous meaning，I'll make some adjustments, thanks！

**Comment by weartist** [comment]
@bpasero I did some modifications and testing, hopefully, all possible scenarios are covered

**Comment by weartist** [comment]
@bpasero I did some modifications and testing, hopefully, all possible scenarios are covered, and It looks like 

https://github.com/microsoft/vscode/blob/ad70ffe4e90f91296407a36286a71e7004b865e0/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L255
and 
https://github.com/microsoft/vscode/blob/ad70ffe4e90f91296407a36286a71e7004b865e0/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L270
can serve the same purpose, I chose handle with the first

File: src/vs/workbench/common/contextkeys.ts [hunk 1]
```diff
@@ -55,6 +55,7 @@ export const EditorGroupEditorsCountContext = new RawContextKey<number>('groupEd
 export const ActiveEditorGroupEmptyContext = new RawContextKey<boolean>('activeEditorGroupEmpty', false, localize('activeEditorGroupEmpty', "Whether the active editor group is empty"));
 export const ActiveEditorGroupIndexContext = new RawContextKey<number>('activeEditorGroupIndex', 0, localize('activeEditorGroupIndex', "The index of the active editor group"));
 export const ActiveEditorGroupLastContext = new RawContextKey<boolean>('activeEditorGroupLast', false, localize('activeEditorGroupLast', "Whether the active editor group is the last group"));
+export const ActiveEditorLastContext = new RawContextKey<boolean>('activeEditorLastContext', false, localize('activeEditorLastContext', "whether the active editor is the last one in its group"));
```
Comment (bpasero) [inline src/vs/workbench/common/contextkeys.ts:58]:
```suggestion
export const ActiveEditorLastContext = new RawContextKey<boolean>('activeEditorLastContext', false, localize('activeEditorLastContext', "Whether the active editor is the last one in its group"));
```

Comment (bpasero) [inline src/vs/workbench/common/contextkeys.ts:58]: Suggest to move this closer to the keys defined above:

```ts
// Editor State Context Keys
...
```

Comment (bpasero) [inline src/vs/workbench/common/contextkeys.ts:58]: Maybe: `ActiveEditorLastInGroupContext`

File: src/vs/workbench/browser/parts/editor/titleControl.ts [hunk 2]
```diff
@@ -91,6 +91,8 @@ export abstract class TitleControl extends Themable {
 
 	private groupLockedContext: IContextKey<boolean>;
 
+	private editorLastContext: IContextKey<boolean>;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/titleControl.ts:94]: Move below `editorPinnedContext` to have them grouped

File: src/vs/workbench/browser/parts/editor/editorGroupView.ts [hunk 3]
```diff
@@ -261,8 +262,11 @@ export class EditorGroupView extends Themable implements IEditorGroupView {
 				activeEditorListener.value = activeEditor.onDidChangeDirty(() => {
 					groupActiveEditorDirtyContext.set(activeEditor.isDirty() && !activeEditor.isSaving());
 				});
+				const indexOfEditor = this.model.indexOf(activeEditor);
+				groupActiveEditorLastContext.set(indexOfEditor + 1 === this.count);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorGroupView.ts:266]:
```suggestion
				groupActiveEditorLastContext.set(this.model.indexOf(activeEditor); + 1 === this.count);
```

**Review by bpasero — CHANGES_REQUESTED** [review]
Besides the :lipstick: feedback, I wonder if the context key fails to update properly when you move a tab, because in that case `onDidActiveEditorChange` will not fire.

**Comment by weartist** [comment]
> Besides the 💄 feedback, I wonder if the context key fails to update properly when you move a tab, because in that case `onDidActiveEditorChange` will not fire.



> Besides the 💄 feedback, I wonder if the context key fails to update properly when you move a tab, because in that case `onDidActiveEditorChange` will not fire.



> I would suggest to look closer at an existing example of a context key that is updated within a group and also for the context menu and just monkey-see-monkey-doo.
> 
> https://github.com/microsoft/vscode/blob/7c1c5300513617d79e1c66c8d2117a5b95259ffa/src/vs/workbench/common/contextkeys.ts#L40
> 
> The context key should not be specific for the context menu, because that is just a corner case.

You are right, we need to handle move events, active events don't handle this well

**Comment by weartist** [comment]
> Besides the 💄 feedback, I wonder if the context key fails to update properly when you move a tab, because in that case `onDidActiveEditorChange` will not fire.

You are right, we need to handle move events, active events don't handle this well, It was an oversight on my part, switch to deal here:
https://github.com/microsoft/vscode/blob/ad70ffe4e90f91296407a36286a71e7004b865e0/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L270

**Comment by bpasero** [comment]
Thanks!

**Pull request merged**